beacon-chain/blockchain/chain_info.go

```diff
@@ -68,19 +70,43 @@ func (s *Service) HeadRoot() []byte {
 
 // HeadBlock returns the head block of the chain.
 func (s *Service) HeadBlock() *ethpb.BeaconBlock {
+	if s.headBlock == nil {
+		head, err := s.beaconDB.HeadBlock(s.ctx)
+		if err != nil {
+			log.Fatalf("Could not get head block: %v", err)
+		}
+		s.headBlock = head
+	}
 	return proto.Clone(s.headBlock).(*ethpb.BeaconBlock)
 }
 
 // HeadState returns the head state of the chain.
 func (s *Service) HeadState() *pb.BeaconState {
+	if s.headState == nil {
+		headState, err := s.beaconDB.HeadState(s.ctx)
+		if err != nil {
+			log.Fatalf("Could not get head state: %v", err)
+		}
+		s.headState = headState
+	}
 	return proto.Clone(s.headState).(*pb.BeaconState)
 }
 
 // CanonicalRoot returns the canonical root of a given slot.
 func (s *Service) CanonicalRoot(slot uint64) []byte {
 	s.canonicalRootsLock.RLock()
 	defer s.canonicalRootsLock.RUnlock()
-
+	if s.canonicalRoots[slot] == nil || len(s.canonicalRoots[slot]) == 0 {
+		head, err := s.beaconDB.HeadBlock(s.ctx)
+		if err != nil {
+			log.Fatalf("Could not get head block: %v", err)
+		}
+		root, err := ssz.SigningRoot(head)
+		if err != nil {
+			log.Fatalf("Could not get hash tree root: %v", err)
+		}
+		s.canonicalRoots[slot] = root[:]
```

Comment:
This is very wrong. Someone could be asking for a head slot of 5 (before) but current head slot is 10. This will assign head root of 10 to head slot of 5.